docs/DeveloperGuide.adoc

```diff
@@ -767,6 +792,229 @@ Use case ends.
+2.  AddressBook shows a list of tasks by tag
+3.  User requests to see the tasks by tag
+4.  AddressBook shows the tasks by tag
++
+Use case ends.
+
+*Extensions*
+
+[none]
+* 2a. The task list is invalid.
++
+[none]
+** 2a1. AddressBook shows an error message.
+** 2a2. AddressBook is empty.
++
+Use case resumes at step 2.
+
+Use case ends.
+
+
+[discrete]
+=== Use case: Add a new task
+
+*MSS*
+
+1.  User requests to add a task
+2.  AddressBook prompts to ask for task details
+3.  User inputs task details
+4.  AddressBook adds the task
++
+Use case ends.
+
+*Extensions*
+
+* 2a. The details is invalid.
++
+[none]
+** 2a1. AddressBook shows an error message.
++
+Use case resumes at step 2.
+
+[discrete]
+=== Use case: Edit hierarchy
+
+*MSS*
+
+1.  User requests to edit hierarchy
+2.  AddressBook prompts to ask for hierarchy details
+3.  User inputs hierarchy details
+4.  AddressBook update hierarchy
++
+Use case ends.
+
+*Extensions*
+
+* 2a. The details is invalid.
++
+[none]
+** 2a1. AddressBook shows an error message.
++
+Use case resumes at step 2.
+
+[discrete]
+=== Use case: View task details
+
+*MSS*
+
+1.  User requests to list tasks
+2.  AddressBook shows a list of persons
+3.  User requests to see a specific task details
+4.  AddressBook shows the details of a person
++
+Use case ends.
+
+*Extensions*
+
+[none]
+* 2a. The list is empty.
++
+Use case ends.
+
+* 3a. The given index is invalid.
++
+[none]
+** 3a1. AddressBook shows an error message.
++
+Use case resumes at step 2.
+[discrete]
+=== Use case: View task details
+
+*MSS*
+
+1.  User requests to list tasks
+2.  AddressBook shows a list of persons
+3.  User requests to see a specific task details
+4.  AddressBook shows the details of a person
++
+Use case ends.
+
+*Extensions*
+
+[none]
+* 2a. The list is empty.
++
+Use case ends.
+
+* 3a. The given index is invalid.
++
+[none]
+** 3a1. AddressBook shows an error message.
++
+Use case resumes at step 2.
+
+[discrete]
+=== Use case: View task details
+
+*MSS*
+
+1.  User requests to list tasks
+2.  AddressBook shows a list of persons
+3.  User requests to see a specific task details
+4.  AddressBook shows the details of a person
++
+Use case ends.
+
+*Extensions*
+
+[none]
+* 2a. The list is empty.
++
+Use case ends.
+
+* 3a. The given index is invalid.
++
+[none]
+** 3a1. AddressBook shows an error message.
++
+Use case resumes at step 2.
+
+[discrete]
+=== Use case: Invalid actions
+
+*MSS*
+
+1.  User requests an invalid action
+2.  AddressBook shows a error message
+3.  AddressBook shows suggestions on fixings of this error or warning occurred
++
+Use case ends.
+
+*Extensions*
+
+[none]
+* 2a. The error message is hard to understand
++
+[none]
+** 2a1. The user can choose to exit/restart the program
+** 2a2. The user can choose to redo this action
++
+Use case ends.
+
+* 3a. The given suggestion is not a good fix
++
+[none]
+** 3a1. AddressBook opt to redo this action
++
+Use case ends.
```

Comment:
If it is an invalid action, how can the app/user redo this action? Error message and and given suggestion is something that you guys have control over, so try to give good ones to avoid the above-mentioned situations :)